### ISSUE TITLE
Support "Zulu OpenJDK Discovery API" for auto provisioning toolchains

### ISSUE THREAD

**Issue by yogurtearl** [gradle/gradle#19140]
Gradle currently supports Adoptium API for auto provisioning. 

This feature request is to also support "Zulu OpenJDK Discovery API" for auto provisioning: 
https://app.swaggerhub.com/apis-docs/azul/zulu-download-community/1.0

when JVM vendor spec is `AZUL("azul systems", "Zulu")` it should autoprovision from the Zulu API.

### Expected Behavior

It generally should operate in the same way as Adoptium auto provisioning.

you should be able to override the base URI of the API with this Gradle property: 
```
org.gradle.jvm.toolchain.install.zulu.baseUri
```

### Context

Zulu supports Apple Silicon (M1 mac) for JDK 8 and JDK 11, Adoptium/Temurin doesn't currently offer builds of JDK 8 and JDK 11 for Apple Silicon (M1 mac), so it would be nice if Gradle supported their API for auto provisioning.

**Comment by ljacomet** [comment]
The plan is to offer an SPI for plugin authors to provide that feature. The migration from AdoptOpenJDK to Adoptium has shown that having this baked into Gradle is a bad idea as these things are outside of our control for their lifecycle.

**Comment by jjcoffman** [comment]
might be worth looking at https://github.com/foojayio/discoapi as this is the Friends of OpenJDK discovery API as well.

**Comment by yschimke** [comment]
This seems like a complicated situation.  The linked API in https://docs.gradle.org/current/userguide/toolchains.html is apparently deprecated.  There are Zulu and Adoptium APIs as well out there, not clear if they conform to the same V1? API.

Is there a matrix of ones that work, Providers/Versions? Or docs for which API baseUris are supported?

The docs make it all sound really hopeful, but I think I've come to the same conclusion of as @aalmiray https://github.com/jreleaser/jreleaser/commit/4362c41e9a3bae09d0af4ad43759b557ff6653e9 that it only actually supports ADOPTOPENJDK.

If I sound frustrated, it's because this seems so close to being really effective solution, and I keep coming back every 3 months and failing.  Any advice that could unblock me is really appreciated!

**Comment by ljacomet** [comment]
Never has Gradle announced that it supported more than AdoptOpenJDK. We made the base URL configurable so that you could decided to support your own server, but that had other side effects.

The team wants to go down the SPI path and will share more information when that topic can be picked up and worked on.

**Comment by yschimke** [comment]
@ljacomet ok thanks, that makes sense.

I read too much optimism into the following, which is presumably written because eventually someone may implement this.  At the moment it seems that these JDK providers are choosing alternative API mechanisms incompatible with Gradle's choice.

> By default, the public AdoptOpenJDK APIs are used to determine and download a matching JDK. In case you want to use another server that is compatible with v3 of the AdoptOpenJDK API, you can point Gradle to use a different host. For that you use the Gradle property as in the example below:

**Comment by CRogers** [comment]
@ljacomet Just wondered if there was any more information on using an SPI to configure JDK provisioning? We're extremely heavy users of Gradle at my compnay and I'm looking at how we can host our jvm discovery server, as calling out to the adoptopenjdk/adoptium api over the internet is not really possible. Our preferred jvm vendor is Azul (we even pay them already) so I'm looking at how to "trick" gradle into downloading a zulu jdk - if there's some work that prevents me from having to doing this (especially if it can be shipped as a gradle plugin) I'd be very interested!

**Comment by ljacomet** [comment]
We are still looking into doing that, no ETA though.

For the moment your only option is to replicate the AdoptOpenJDK URL structure on a web server and serve Azul JDKs from it, without using the vendor in the toolchain spec.

**Comment by yogurtearl** [comment]
Also, existing AdoptOpenJdk auto-provisioning is not working for JDK 18: 
https://github.com/gradle/gradle/issues/20344

**Comment by CRogers** [comment]
For people's SA, we've ended up [building our own plugin for JDK auto-provisioning](https://github.com/palantir/gradle-jdks) and started using it on all our projects.

[Features over the existing auto-provisioning](https://github.com/palantir/gradle-jdks#motivation) include:

- You can choose your favoured JDK vendor.
- Use the same JDK version for representative, reproducible builds on dev machines and in CI.
- Automatically add JDK CA certificates.
- Point to an internal mirror for JDKs.

**Comment by tomakita** [comment]
@CRogers your plugin is awesome -- thank you for sharing it!  

Out of curiosity, did you consider submitting this functionality as a Gradle PR (to implement https://github.com/gradle/build-tool-roadmap/issues/29), instead of as a standalone plugin?  We are considering creating something similar, and were considering submitting a PR for https://github.com/gradle/build-tool-roadmap/issues/29, when we found your plugin.  We're thinking it would be ideal to have this functionality as a first class part of Gradle Toolchains, because eventually, IDEs could integrate with Gradle Toolchains and use whatever JDK your Toolchain is configured to use (so you get the correct code completion and syntax highlighting, for example).

**Comment by SamB** [comment]
@tomakita Well, somebody tried to implement something of that nature over at https://github.com/gradle/gradle/pull/18028, but it's not clear where the review ended up. Getting a PR actually merged isn't totally trivial...

**Comment by ataverasp** [comment]
for paying customers of Zulu we need to provide credentials would this API accept credentials for auto provisioning with Zulu?

**Comment by ljacomet** [comment]
Conceptually, what we are planning is to have an SPI so that plugin authors can design their own logic for matching a toolchain spec to a URI.
Plugin configuration will allow to pass in credentials if the endpoint requires them.

Will try to share an example soon!

**Comment by aalmiray** [comment]
Any links to documentation on how to use this feature now that it has been marked as completed?

**Comment by sschuberth** [comment]
> Any links to documentation on how to use this feature now that it has been marked as completed?

There's some docs at https://github.com/gradle/gradle/pull/22123/files#diff-23ed1cf4b4d7e28f4236ae239e1a4546983788d249ac320fec08642240ff3b43R272.

**Comment by ljacomet** [comment]
[Release notes for nightlies](https://docs.gradle.org/release-nightly/release-notes.html#added-support-for-java-toolchain-downloads-from-arbitrary-repositories) can be used as the entry point.
We will also release a plugin close to the effective 7.6 release.

**Comment by jbartok** [comment]
https://docs.gradle.org/release-nightly/userguide/toolchains.html#sub:download_repositories

**Comment by yogurtearl** [comment]
> We will also release a plugin close to the effective 7.6 release.

@ljacomet is this plugin released? is the in-progress source code for that plugin available anywhere? 

the docs/release notes don't make it clear that `AzulResolver` is from a theoretical future plugin, and not included in Gradle 7.6. Might be worth updating the release notes, to avoid people copy/pasting the `AzulResolver` stuff and then filing issues saying it doesn't work. :) 

https://docs.gradle.org/current/release-notes.html#toolchain-provision

**Comment by yogurtearl** [comment]
[As mentioned above](https://github.com/gradle/gradle/issues/19140#issuecomment-989221915), would be great to have support for the [foojay Disco API
](https://github.com/foojayio/discoapi) as well (but maybe that is a separate ticket).

**Comment by sschuberth** [comment]
> the docs/release notes don't make it clear that AzulResolver is from a theoretical future plugin, and not included in Gradle 7.6

@yogurtearl, I'd say that

> For example, the following uses *custom plugins* that provide AzulResolver and AdoptiumResolver to add custom toolchains for Adoptium and Azul:

as written [here](https://docs.gradle.org/current/release-notes.html#toolchain-provision) (emphasis mine) is pretty clear about that.

**Comment by sschuberth** [comment]
> would be great to have support for the [foojay Disco API](https://github.com/foojayio/discoapi)

You mean like in https://github.com/gradle/foojay-toolchains? :wink:
